core/src/main/java/com/novoda/noplayer/internal/exoplayer/mediasource/MediaSourceFactory.java

```diff
@@ -43,19 +42,18 @@ public MediaSourceFactory(Context context,
 
     public MediaSource create(Options options,
                               Uri uri,
-                              MediaSourceEventListener mediaSourceEventListener,
-                              DefaultBandwidthMeter bandwidthMeter,
-                              Optional<AdsLoader> advertsLoader) {
+                              ExoPlayerListener exoPlayerListener,
+                              DefaultBandwidthMeter bandwidthMeter) {
         DefaultDataSourceFactory defaultDataSourceFactory = createDataSourceFactory(bandwidthMeter);
 
         MediaSource contentMediaSource = getMediaSourceFor(options, uri, defaultDataSourceFactory);
 
-        if (advertsLoader.isPresent()) {
-            AdsMediaSource adsMediaSource = new AdsMediaSource(contentMediaSource, defaultDataSourceFactory, advertsLoader.get(), null);
-            adsMediaSource.addEventListener(handler, mediaSourceEventListener);
+        if (exoPlayerListener.canLoadAdverts()) {
```

Comment:
This is looking weird to me, as I wouldn't expect a listener (by definition passive) to dictate changes in behaviour here.